source/unix/openssl_aes.c

```diff
@@ -0,0 +1,717 @@
+    .finalize_encryption = s_finalize_encryption,
+};
+
+struct aws_symmetric_cipher *aws_aes_ctr_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv) {
+    struct openssl_aes_cipher *cipher = aws_mem_calloc(allocator, 1, sizeof(struct openssl_aes_cipher));
+
+    cipher->cipher_base.allocator = allocator;
+    cipher->cipher_base.block_size = AWS_AES_256_CIPHER_BLOCK_SIZE;
+    cipher->cipher_base.key_length_bits = AWS_AES_256_KEY_BIT_LEN;
+    cipher->cipher_base.vtable = &s_ctr_vtable;
+    cipher->cipher_base.impl = cipher;
+
+    if (key) {
+        aws_byte_buf_init_copy_from_cursor(&cipher->cipher_base.key, allocator, *key);
+    } else {
+        aws_byte_buf_init(&cipher->cipher_base.key, allocator, AWS_AES_256_KEY_BYTE_LEN);
+        aws_symmetric_cipher_generate_key(AWS_AES_256_KEY_BYTE_LEN, &cipher->cipher_base.key);
+    }
+
+    if (iv) {
+        aws_byte_buf_init_copy_from_cursor(&cipher->cipher_base.iv, allocator, *iv);
+    } else {
+        aws_byte_buf_init(&cipher->cipher_base.iv, allocator, AWS_AES_256_CIPHER_BLOCK_SIZE);
+        aws_symmetric_cipher_generate_initialization_vector(
+            AWS_AES_256_CIPHER_BLOCK_SIZE, false, &cipher->cipher_base.iv);
+    }
+
+    /* EVP_CIPHER_CTX_init() will be called inside EVP_CIPHER_CTX_new(). */
+    cipher->encryptor_ctx = EVP_CIPHER_CTX_new();
+    AWS_FATAL_ASSERT(cipher->encryptor_ctx && "Cipher initialization failed!");
+
+    /* EVP_CIPHER_CTX_init() will be called inside EVP_CIPHER_CTX_new(). */
+    cipher->decryptor_ctx = EVP_CIPHER_CTX_new();
+    AWS_FATAL_ASSERT(cipher->decryptor_ctx && "Cipher initialization failed!");
+
+    if (s_init_ctr_cipher_materials(&cipher->cipher_base) != AWS_OP_SUCCESS) {
+        goto error;
+    }
+
+    cipher->cipher_base.good = true;
+    return &cipher->cipher_base;
+
+error:
+    s_destroy(&cipher->cipher_base);
+    return NULL;
+}
+
+static int s_finalize_gcm_encryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out) {
+    struct openssl_aes_cipher *openssl_cipher = cipher->impl;
+
+    int ret_val = s_finalize_encryption(cipher, out);
+
+    if (ret_val == AWS_OP_SUCCESS) {
+        if (!cipher->tag.len) {
+            if (!EVP_CIPHER_CTX_ctrl(
+                    openssl_cipher->encryptor_ctx,
+                    EVP_CTRL_GCM_GET_TAG,
+                    (int)cipher->tag.capacity,
+                    cipher->tag.buffer)) {
+                cipher->good = false;
+                return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+            }
+            cipher->tag.len = AWS_AES_256_CIPHER_BLOCK_SIZE;
+        }
+    }
+
+    return ret_val;
+}
+
+static int s_init_gcm_cipher_materials(struct aws_symmetric_cipher *cipher) {
+    struct openssl_aes_cipher *openssl_cipher = cipher->impl;
+
+    if (!(EVP_EncryptInit_ex(openssl_cipher->encryptor_ctx, EVP_aes_256_gcm(), NULL, NULL, NULL) &&
+          EVP_EncryptInit_ex(
```

Comment:
is there something specific about gcm that requires splitting init into 2 calls? i dont see openssl docs mentioning the need for it

Reply:
yes, it's openssl API across multiple versions protection. This is the code that we've been able to work across all versions.